File: app/models/domain/report/Report.scala
package models.domain.report

import models.dao.ReportDAO
import models.domain.util.Date
import models.ebean.{Report => EReport}
import play.api.libs.json.{Json, OFormat}

case class Report(id: Option[Long], userId: Long, description: String, date: Date, solved: Boolean, rejected: Boolean) {
  def equals(report: Report): Boolean = {
    if(report.id.isDefined && id.isDefined) id.get.equals(report.id.get)
    else false
  }
}

object Report extends ReportJsonFormat {

  def apply(report: EReport): Report = {
    Report(
      Option(report.getId),
      report.getUser.getId,
      report.getDescription,
      Date(report.getDate),
      report.getSolved,
      report.getRejected
    )
  }

  def apply(reportCreate: ReportCreate): Report = {
    Report(
      None,
      reportCreate.userId,
      reportCreate.description,
      Date.now,
      false,
      false
    )
  }

  def saveOrUpdate(report: Report): Option[Report] = {
    ReportDAO.saveOrUpdate(report)
  }

  def getById(id : Long) : Option[Report] = {
    ReportDAO.getById(id)
  }

  def getByUserId(id : Long) : List[Report] = {
    ReportDAO.getByUserId(id)
  }

  def getAll: List[Report] = {
    ReportDAO.getAllReports
  }

  def reject(id: Long): Boolean = {
    ReportDAO.reject(id)
  }

  def getUnresolved: List[Report] = {
    ReportDAO.getReportsUnresolved
  }

  def delete(report: Report): Option[Boolean] = {
    ReportDAO.delete(report)
  }

  def getUserReportStatistics(id: Long): ReportStatistics = {
    ReportStatistics(ReportDAO.getNumberOfTotalUserReports(id).longValue(), ReportDAO.getNumberOfRejectedUserReports(id).longValue(), ReportDAO.getNumberOfSolvedUserReports(id).longValue())
  }

}

trait ReportJsonFormat{
  implicit val reportFormat: OFormat[Report] = Json.format[Report]
}

File: app/models/ebean/Report.java
package models.ebean;

import io.ebean.Ebean;
import io.ebean.Finder;
import io.ebean.Model;
import io.ebean.annotation.NotNull;
import org.joda.time.DateTime;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import java.util.List;
import java.util.Optional;

@Entity
public class Report extends Model {

    @Id
    @NotNull
    private Long id;
    @NotNull
    @ManyToOne
    private User user;
    private String description;
    private DateTime date;
    private Boolean solved;
    private Boolean rejected;


    private static Finder<Long, Report> finder = new Finder<>(Report.class);

    public Report(Long id, User user, String description, DateTime date, Boolean solved, Boolean rejected) {
        this.id = id;
        this.date = date;
        this.description = description;
        this.user = user;
        this.solved = solved;
        this.rejected = rejected;
    }

    @Override
    public void save(){
        super.save();
    }

    public static Optional<Report> getReportById(Long id){
        Report report = finder.byId(id);
        if(report != null){
            return Optional.of(report);
        }
        return Optional.empty();
    }

    public static List<Report> getReportByUserId(Long id){
        return Ebean.find(Report.class).where().eq("user_id", id).findList();
    }

    public static List<Report> getReportsUnsolved(){
        return Ebean.find(Report.class).where().eq("solved", 0).eq("rejected", 0).findList();
    }

    public static Integer getNumberOfRejectedUserReports(Long id){
        return Ebean.find(Report.class).where().eq("user_id", id).eq("rejected", 1).findCount();
    }

    public static Integer getNumberOfSolvedUserReports(Long id){
        return Ebean.find(Report.class).where().eq("user_id", id).eq("solved", 1).findCount();
    }

    public static Integer getNumberOfTotalUserReports(Long id){
        return Ebean.find(Report.class).where().eq("user_id", id).findCount();
    }

    public static List<Report> getAllReports() {
        return finder.all();
    }

    public Long getId() {
        return id;
    }

    public String getDescription() {
        return description;
    }

    public DateTime getDate() {
        return date;
    }

    public Boolean getSolved() {
        return solved;
    }

    public User getUser() {
        return user;
    }

    public Boolean getRejected() {
        return rejected;
    }
}
